In [7]:
import math
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as stats

#### 2.1 - Discrete random variable

##### a) Let X be a stochastic variable. When running the Python-command `stats.binom.pmf(4, 10, 0.6)` Python returns 0.1115. written as:

`print(stats.binom.pmf(4, 10, 0.6))` is 0.11147673600000009

*What distribution is applied and what does 0.1115 represent?*

Important note: Since they are using .pmf on the binomial distribution, does it represent the exact probability of P(X = x). (gives you the probability of exactly 
X successes.)

We see from the python command that the **binomial distribution** is being used. This distrabution is often used, when you have two outcome success or failure. 

From the commando, we can analyse that x = 4, n = 10 and p = 0.6 (60%).

**0.1115 represents the probability to get 4 successes in 10 tries with a properbility of 60%** 

##### b) Let X be the same stochastic variable as above. The following are results from Python:

`print(stats.binom.cdf(4,10,0.6))` is 0.16623861760000005

`print(stats.binom.cdf(5,10,0.6))` is 0.3668967424000001

*Calculate the following probabilities: P(X ≤ 5), P(X < 5), P(X > 4) and P(X = 5).*

Important note: Since they use .cdf on the binomial distribution, the calculations represent the probability P(X <= x) (gives you the probability of 
X or fewer successes.)

P(X <= 5) is the probability of getting 5 or less success draws from 10 draws with a properbility of 60%. This is already calculated for us with the second python command from the exercise.

Therefore, $$P(X <= 5) = 0.3668967424000001$$

P(X > 4) is the probability of getting more than 4 success draws from 10 draws (p=0.6). Thankfully they have given us the probability to draw 4 or less successes. We can then do,

$$P(X > 4) = 1 - P(X <= 4) = 1 - 0.16623861760000005 = 0.83376$$

This is the result from the first python calculation:
$$P(X < 5) = 0.16623861760000005$$

To find the exact probability for P(X = 5), we do the following calculation:
$$P(X = 5) = P(X <= 5) - P(X <= 4) =  0.3668967424000001 - 0.16623861760000005 = 0.20065$$



##### c) Let X be a stochastic variable. From Python we get:

`print(stats.poisson.pmf(k=4,mu=3))` is 0.16803135574154085

*What distribution is applied and what does 0.16803 represent?*

By looking at the Python commando, can we see that the distribution is the Poisson distribution.

The Poisson distribution describes the probability of a given number of events occurring in a fixed interval (distance, area or volume).

λ(mu) = 3, is the rate or intensity: The average number of events per interval. 

k = 4, is the change of getting exactly (because we use the pmf) 4 emails in a given hour (interval).

**So P(4; 3) = 0.16803135574154085 is the probability of getting, lets say 4 emails in a given hour.**

##### d) Let X be the same stochastic variable as above. The following are results from Python:

`print(stats.poisson.cdf(4,3))` = 0.8152632445237722

`print(stats.poisson.cdf(5,3))` = 0.9160820579686966

*Calculate the following probabilities: P(X <= 5), P(X < 5), P(X > 4) and P(X = 5)*

Important note. We are using the cdf on the poisson destribution, which represent the probability of getting X or fewer.

These we just read from the results above from the python commands:

$$P(X <= 5) = 0.9160820579686966$$
$$P(X < 5) = 0.8152632445237722$$

P(X < 5) represent 1,2,3,4. We subtract that with 1, and we get the probability for X > 4.

$$P(X > 4) = 1 - P(X < 5) = 1 - 0.8152632445237722 = 0.184736755$$


P(X <= 5) represent 1,2,3,4,5, while P(X < 5) represent 1,2,3,4. Subtract these and you end up with only the probability of 5 P(X = 5).

$$P(X = 5) = P(X <= 5) - P(X < 5) = 0.9160820579686966 - 0.8152632445237722 = 0.10081881344492438$$

#### 2.4 - Consumer survey

In a consumer survey performed by a newspaper, 20 different groceries (products) were purchased in a grocery store. Discrepancies between the price appearing on the sales slip and the shelf price were found in 6 of these purchased
products.

##### a) At the same time a customer buys 3 random (different) products within the group consisting of the 20 goods in the store. The probability that no discrepancies occurs for this customer is?

First we analyze what distribution we would use for this scenario. The Poission Distribution would not make sense here, as that used to represent probability over intervals and here we have none. It would also not make sense to use the Binomial distribution as the customer picks 3 **different** products, which means without replacement. We therefore conclude to use the Hypergeometric Distribution.

- n = Number of draws,
- a = Number of successes in population,
- N = Population Size,
- x = Number of successes in sample.

$$f(x;n,a,N) = P(X = x) = \frac{\binom{a}{x} \binom{N - a}{n - x}}{\binom{N}{n}}$$

$$\binom{a}{b} = \frac{a!}{b!(a-b)!'}$$

The customer buy 3 products (n = 3). We have a total of 20 different products (N = 20). We have 6 discrepancies prices found (a = 6). We want to calculate the probability for the customer to have bought 0 discrepanies products (x = 0). We will insert into the formula with our observations.

$$f(x;n,a,N) = P(X = 0) = \frac{\binom{6}{0} \binom{20 - 6}{3 - 0}}{\binom{20}{3}} = \frac{\binom{6}{0} \binom{14}{3}}{\binom{20}{3}} \leftrightarrow$$

Calculating the different binoms with the formula above:

$$\binom{6}{0} = \frac{6!}{0! (6-0)!} = \frac{6!}{6!} = 1$$

$$\binom{14}{3} = \frac{14!}{3! (14 - 3)!} = \frac{14!}{3! \cdot 11!} = \frac{87.178.291.200}{6 \cdot 39.916.800} = \frac{87.178.291.200}{239.500.800} = 364$$

$$\binom{20}{3} = \frac{20!}{3! (20 - 3)!} = \frac{20!}{3! \cdot 17!} = \frac{2.432.902.008.000.000.000}{6 \cdot 355.687.428.100.000} = \frac{2.432.902.008.000.000.000}{2.134.124.569.000.000} = 1140$$

Insert the binoms inside the formula and we get

$$f(x;n,a,N) = P(X = 0) = \frac{\binom{6}{0} \binom{14}{3}}{\binom{20}{3}} = \frac{1 \cdot 364}{1140} = \frac{364}{1140} = \underline{\underline{\frac{91}{285} \approx 0.3192}}$$

We conclude that the probability for the customer to have bought 0 discrepanies products are 0.3192 * 100 = 31.92%


#### 2.5 - Hay delivery quality

A horse owner receives 20 bales of hay in a sealed plastic packaging. To control the hay, 3 bales of hay are randomly selected, and each checked whether it contains harmful fungal spores.

It is believed that among the 20 bales of hay 2 bales are infected with fungal spores. A random variable X describes the number of infected bales of hay among the three selected.

##### a) The mean of X ($\mu_{x}$), the variance of X $(\sigma^{2}_{x})$ and P(X >= 1) are?

To figure out the formulas for the mean and variance, we need to determine what distribution is the correct one to use for this math problem. Again the Poisson distribution is not the correct one to use, because our problem has nothing to do with intervals over time. We then look at the Binomial distribution or Hypergeometric distribution. Since we have without replacement in this problem, we pick 3 randomly selected haybales and therefore will not check them again, we have to use the Hypergeometric distribution. 

See 2.4 for how the hypergeometric distribution looks.

We then analyse from the math problem that we have a population of 20 haybales (N = 20). We pick 3 random selected haybales (n = 3). The number of infected hay bales are 2 (a = 2).

Summary: N = 20, n = 3, a = 2 and x = 0 (will explain later).

To calculate the mean and variance are we using the following formula for Hypergemo:

Mean:

$$\mu = n \frac{a}{N} = 3 \cdot \frac{2}{20} = \frac{3 \cdot 2}{20} = \frac{6}{20} = \underline{\underline{\frac{3}{10} \approx 0.3}}$$

Variance:

$$
\begin{aligned}
\sigma^2 = n \frac{a(N-a)}{N^2} \frac{N - n}{N - 1} = 3 \cdot \frac{2 (20-2)}{20^2} \cdot \frac{20-3}{20-1} &= \\ \\
3 \cdot \frac{2 \cdot 18}{400} \cdot \frac{20-3}{20-1} = 3 \cdot \frac{36}{400} \cdot \frac{20-3}{20-1} &= \\ \\
\frac{3 \cdot 36}{400} \cdot \frac{20-3}{20-1} = \frac{108}{400} \cdot \frac{20-3}{20-1} &= \\ \\
\frac{108}{400} \cdot \frac{17}{19} = 0.27 \cdot 0.8947 &= \underline{\underline{0.2415}}
\end{aligned}
$$

To calculate the probability that there is one or more infected haybales, we first have to calculate the probability that no haybales have been infeted P(X = 0). 

Thats why we set x = 0 earlier. With this probability of no infested haybales, we cal calculate P(X >= 1) = 1 - P(x = 0).

We start of by inserting the values we gathered from the math problem:

$$P(X=0) = \frac{\binom{2}{0} \binom{20-2}{3-0}}{\binom{20}{3}} = \frac{\binom{2}{0} \binom{18}{3}}{\binom{20}{3}}$$

We then calculate each binom. Note: I am not including every intermediate calculation for this. See exercise 2.4 for deeper dive.

$$\binom{2}{0} = \frac{2!}{0!(2-0)!} = \frac{2!}{2!} = 1$$

$$\binom{18}{3} = \frac{18!}{3!(18-3)!} = \frac{18!}{3! \cdot 15!} = 816$$

$$\binom{20}{3} = \frac{20!}{3!(20-3)!} = \frac{20!}{3! \cdot 17!} = 1140$$

Insert the calculated binoms:

$$P(X=0) = \frac{\binom{2}{0} \binom{18}{3}}{\binom{20}{3}} = \frac{1 \cdot 816}{1140} = \frac{816}{1140} = 0.7157$$

We can then calculate P(X >= 1) as follows

$$P(X >= 1) = 1 - P(X=0) = 1 - 0.7157 = \underline{\underline{0.284}}$$

We conclude that there is a $0.284 \cdot 100 = 28.4%$ chance that there is 1 or more haybales infected with harmful fungal spores

In [8]:
# How to calculate it in Python.

# Init the values from the exercise
N = 20
n = 3
a = 2
x = 0

# Calculate mean stats.hypergeom.mean(population, successes, draws)
mean = stats.hypergeom.mean(N, a, n)
print(f'Hypergeom Mean: {mean}')

# Calculate variance stats.hypergeom.var(population, successes, draws)
variance = stats.hypergeom.var(N, a, n)
print(f'Hypergeom Variance: {variance}')

# Calculate Probability P(X >= 1)

# P(X = 0)
prob_a = stats.hypergeom.pmf(x, N, a, n)

# P(x >= 1)
prob_b = (1 - prob_a)

print(f'P(X >= 1): {prob_b * 100}')

Hypergeom Mean: 0.3
Hypergeom Variance: 0.24157894736842106
P(X >= 1): 28.421052631578945


##### b) Another supplier advertises that no more than 1% of his bales of hay are infected. The horse owner buys 10 bales of hay from this supplier, and decides to buy hay for the rest of the season from this supplier if the 10 bales are error-free. 

*What is the probability that the 10 purchased bales of hay are error-free, if 1% of the bales from a supplier are infected (p1) and the probability that the 10 purchased bales of hay are error-free, if 10% of the bales from a supplier are infected (p10)?*

This task includes multiple questions we have to answer. 

- What is the probability that the 10 hay bales are error-free, if 1% of the bales from a supplier are infected (p1)
- What is the probability that the 10 hay bales are error-free, if 10% of the bales from the supplier are infected (p10)

p1:

Since we do not know the size of the suppliers haybals, we then have to use the Binomial Distribution. See the sentence below

"Sampling with replacement" = "Sampling from an infinite population"

The Binomial Distribution looks like this:

$$P(X = x) = \binom{n}{x} p^x (1-p)^{n-x}$$

$$\binom{n}{x} = \frac{n!}{x!(n-x)!}$$

From reading the text, we get the following n and p, where n is the number of draws and p is the probability of infected haybal. We set x = 0, because we want to calculate the probability of all 10 hayballs not being infected P(X = 0).

n = 10,

p = 0.01 (1%), 

x = 0

Insert the values into the formula for Binomial Distribution:

$$P(X = 0) = \binom{10}{0} \cdot 0.01^0 (1-0.01)^{10-0} =$$

Calculate the binom

$$\binom{10}{0} = \frac{10!}{0! (10 - 0)!} = 1$$

Insert the results and calculates to the end. I am not including more details intermediate calculations, as these are done through the allowed calculator and paper.

$$
\begin{aligned}
P(X = 0) = 1 \cdot 0.01^0 (1-0.01)^{10-0} &= \\ \\
1 \cdot 1 (1-0.01)^{10-0} &= \\ \\
1 \cdot 1 \cdot 0.9043 &= 0.9043
\end{aligned}
$$

The probability of 0 of the 10 haybals to be infected is 0.9043 * 100 = 90.43%

p10:

We know do the exact same calculation for probability of infection of 10%

n = 10, p = 0.10 (10%), x = 0

$$P(X = 0) = \binom{10}{0} \cdot 0.10^0 (1-0.10)^{10-0} =$$

$$\binom{10}{0} = \frac{10!}{0! (10 - 0)!} = 1$$

$$
\begin{aligned}
P(X = 0) = 1 \cdot 0.10^0 (1-0.01)^{10-0} &= \\ \\
1 \cdot 1 (1-0.10)^{10-0} &= \\ \\
1 \cdot 1 \cdot 0.3486 &= 0.3486
\end{aligned}
$$

The probability of 0 of the 10 haybals to be infected is 0.3486 * 100 = 34.86%

Alternative approch to both calculations. You could also calculate them by just saying

$$P(X = 0) = (1 - p)^n$$

$$p_1 = P(X_1 = 0) = (1 - 0.01)^10 = 0.99^{10} = 0.9044$$
$$p_10 = P(X_10 = 0) = (1 - 0.10)^10 = 0.90^{10} = 0.3487$$

I would personally use the Binon Distribution, because this is a one of example of a alternative way to calculate it

In [9]:
# Using the power of Python to calculate this scenario for us. We can use both pmf and cdf to come to the same conclusion

# P1 pmf (aslo known as pdf from the book)
p1_pmf = stats.binom.pmf(0, n=10, p=0.01)
p1_cdf = stats.binom.cdf(0, n=10, p=0.01)

print(f'p1 - pmf: {p1_pmf}, cdf: {p1_cdf}')

# P10 
p10_pmf = stats.binom.pmf(0, n=10, p=0.10)
p10_cdf = stats.binom.cdf(0, n=10, p=0.10)

print(f'p1 - pmf: {p10_pmf}, cdf: {p10_cdf}')


p1 - pmf: 0.9043820750088041, cdf: 0.9043820750088045
p1 - pmf: 0.34867844009999993, cdf: 0.34867844009999993


#### 2.7 - A fully automated production

On a large fully automated production plant items are pushed to a side band at random time points, from which they are automatically fed to a control unit. The production plant is set up in such a way that the number of items sent to
the control unit on average is 1.6 item pr. minute. Let the random variable X denote the number of items pushed to the side band in 1 minute. It is assumed that X follows a Poisson distribution.

##### a) What is the probability that there will arrive more than 5 items at the control unit in a given minute is?

Since we are working with a time interval we will use the Poisson Distribution that have the following formula for P(X = x)

$$f(x; \lambda) = P(X = x) = \frac{\lambda^x}{x!} e^{-\lambda}$$

where lambda is the time interval (rate / intensity)

Reading from the text, we get the following variables:

$$\lambda = 1.6$$

To calculate the probability that there will arrive more than 5 items at the control unit in a given minute is

P(X > 5) = 1 - P(X >= 5)

P(X <= 5) = P(X = 0) + P(X = 1) + P(X = 2) + P(X = 3) + P(X = 4) + P(X = 5)

I am only showing detail calculations for P(X = 1). The rest is done through the calculator

$$
\begin{aligned}
P(X = 0) = \frac{1.6^0}{0!} e^{-1.6} &= 0.2018965 \\ \\
P(X = 1) = \frac{1.6^1}{1!} e^{-1.6} &= \frac{1.6}{1} e^{-1.6} = 1.6 \cdot e^{-1.6} = 1.6 0.201896 = 0.32303 \\ \\
P(X = 2) = \frac{1.6^2}{2!} e^{-1.6} &= 0.258427 \\ \\
P(X = 3) = \frac{1.6^3}{3!} e^{-1.6} &= 0.137828 \\ \\
P(X = 4) = \frac{1.6^4}{4!} e^{-1.6} &= 0.0551312 \\ \\
P(X = 5) = \frac{1.6^5}{5!} e^{-1.6} &= 0.017641
\end{aligned}
$$

Calculating the P(X >= 5)

$$P(X >= 5) = 0.2018965 + 0.32303 + 0.258427 + 0.137828 + 0.0551312 + 0.017641 = 0.994$$

Calculating P(X > 5)

$$P(X > 5) = 1 - P(X <= 5) = 1 - 0.994 = 0.006$$

There is a approximately 0.6% (0.006 * 100 = 0.6) probability that there arrive 5 or more items at the control unit in a given minute

In [10]:
# How to calculate it in Python
print(1-stats.poisson.cdf(k=5,mu=1.6))

0.006040291111581331


##### b) What is the probability that no more than 8 items arrive to the control unit within a 5-minute period?

We get rate 1.6 per minut. We need to calculate the rate for 5 minutes. We can do this by multipling by 5 like this. 

$$\lambda_{1} = 1.6$$

$$\lambda_{5} = 1.6 \cdot 5 = 8$$

To find the probability that no more than 8 items arrive, we need to calculate the probability P(X <= 8)

$$P(X <= 8) = P(X = 0) + P(X = 1) + P(X = 2) + P(X = 3) + P(X = 4) + P(X = 5) + P(X = 6) + P(X = 7) + P(X = 8)$$

Calculates all the possiblies

$$
\begin{aligned}
P(X = 0) = \frac{8^0}{0!} e^{-8} &= 0.000335 \\ \\
P(X = 1) = \frac{8^1}{1!} e^{-8} &= 0.002683 \\ \\
P(X = 2) = \frac{8^2}{2!} e^{-8} &= 0.010734 \\ \\
P(X = 3) = \frac{8^3}{3!} e^{-8} &= 0.028626 \\ \\
P(X = 4) = \frac{8^4}{4!} e^{-8} &= 0.057252 \\ \\
P(X = 5) = \frac{8^5}{5!} e^{-8} &= 0.091603 \\ \\
P(X = 6) = \frac{8^6}{6!} e^{-8} &= 0.122138 \\ \\
P(X = 7) = \frac{8^7}{7!} e^{-8} &= 0.139586 \\ \\
P(X = 8) = \frac{8^8}{8!} e^{-8} &= 0.139586 \\ \\
\end{aligned}
$$

Combines them

$$P(X <= 8) = 0.000335 + 0.002683 + 0.010734 + 0.028626 + 0.057252 + 0.091603 + 0.122138 + 0.139586 + 0.139586 = 0.593$$

We conclude that the probability for no more than 8 items to arrive within 5 min is 0.593 * 100 = 59.3%

#### 2.8 - Call center staff

The staffing for answering calls in a company is based on that there will be 180 phone calls per hour randomly distributed. If there are 20 calls or more in a period of 5 minutes the capacity is exceeded, and there will be an unwanted waiting time, hence there is a capacity of 19 calls per 5 minutes.

##### a) What is the probability that the capacity is exceeded in a random period of 5 minutes?

Nothing from this task is different from the task above 2.7b. The same methods are applied, but the analyse of the problem is a bit different.

We use the Poisson Distribution since we're working with time. We get a rate for 180 calls an hour and max capacity of 19 calls per 5 min.

Calculate rate for 5 min. Finding rate for a min by dividing with 60 and then calculate the rate for 5 min by miltiplying with 5.

$$\lambda_{1hour} = 180$$
$$\lambda_{1min} = \frac{180}{60} = 3$$
$$\lambda_{5min} = 3 \cdot 5 = 15$$

We than have to calculate the probability of more calls than 19 in a timespan of 5 min. We express this with the following:

$$P(X > 19) = 1 - P(X <= 19)$$

We know from the previous task that the calculation of P(X <= x) is the PDF combined of probability P(X = 0), P(X = 1)... etc to P(X = 19).

Since we have a great example of doing this for exercise 2.7b, I will reference to that exercise and then skip the manuel calculation of this. I have to calculate 19 probabilities.

We find the result with programmel below instead.

We conclude that there is a 12.47% probability that there will be more than 19 calls within a time period of 5 min.

In [15]:
# The Power of Python Programmel

print(f'P(X > 19): {1-stats.poisson.cdf(k=19, mu=15)}, In Procent: {(1-stats.poisson.cdf(k=19, mu=15)) * 100}%')

print(stats.poisson.cdf(k=19, mu=15))

P(X > 19): 0.12478121503252493, In Procent: 12.478121503252492%
0.8752187849674751


##### b) If the probability should be at least 99% that all calls will be handled without waiting time for a randomly selected period of 5 minutes, how large should the capacity per 5 minutes then at least be?

The current threadshold covers 0.875218 (87.5%) of all calls with x = 19. We know this by calculating P(X <= 19)

$$P(X <= 19) = 1 - P(X > 19) = 0.875218$$

To find the number of x we need to increase, we need to calculate the probability that P(X <= 20) covers and increase x until we find a x that covers 0.99 (99%). 

I will now show my calculations until I have found the x and conclude. This can be done much quicker on programmel, but I will show how to do that after:

$$
\begin{aligned}
P(X <= 19) &= P(X > 19) - 1 = 0.875218 \\ \\
P(X = 20) &= \frac{15^20}{20!} \cdot e^{-15} = 0.041810 \\ \\
P(X <= 20) &= P(X > 19) + P(X = 20) = 0.875218 + 0.041810 = 0.917028 \\ \\
P(X = 21) &= \frac{15^21}{21!} \cdot e^{-15} = 0.029864 \\ \\
P(X <= 21) &= P(X <= 20) + P(X = 21) = 0.917028 + 0.029864 = 0.946892 \\ \\
P(X = 22) &= \frac{15^22}{22!} \cdot e^{-15} = 0.020362 \\ \\
P(X <= 22) &= P(X <= 21) + P(X = 22) = 0.946892 + 0.020362 = 0.967254 \\ \\
P(X = 23) &= \frac{15^23}{23!} \cdot e^{-15} = 0.013279 \\ \\
P(X <= 23) &= P(X <= 22) + P(X = 23) = 0.967254 + 0.013279 = 0.980533 \\ \\
P(X = 24) &= \frac{15^24}{24!} \cdot e^{-15} = 0.008299 \\ \\
P(X <= 24) &= P(X <= 23) + P(X = 24) = 0.980533 + 0.008299 = 0.9888832 \\ \\
(X = 25) &= \frac{15^25}{25!} \cdot e^{-15} = 0.004979 \\ \\
P(X <= 25) &= P(X <= 24) + P(X = 25) = 0.9888832 + 0.004979 = 0.993811 \\ \\
\end{aligned}
$$

The threadshold then need to be x = 25 to cover 0.99 (99%) of the calls coming within a 5 min time interval.

As you see from the calculations is that its a time consuming job to just keep calculating and increasing x until 99% is convered. In programmel, we can do this much easier by using the ppf to find the 99 procentile

In [16]:
# The Power Pythoooon. We have two methods to find the answer. Calculate like we did, or use the ppf to find the 99 procentile

# Manuel calculation and analyzing
print(stats.poisson.cdf(k=[22,23,24,25,26], mu=15))

# Find with the 99 procentile directly
print(stats.poisson.ppf(q=0.99, mu=15))

[0.96725576 0.98053543 0.98883522 0.9938151  0.9966881 ]
25.0
